pkg/dequeuer/batch_handler.go

```diff
@@ -0,0 +1,226 @@
+	"github.com/xtgo/uuid"
+	"go.uber.org/zap"
+)
+
+const (
+	// CortexJobIDHeader is the header containing the job id for the user container
+	CortexJobIDHeader        = "X-Cortex-Job-ID"
+	_jobCompleteMessageDelay = 10 * time.Second
+)
+
+type BatchMessageHandler struct {
+	config                  BatchMessageHandlerConfig
+	jobCompleteMessageDelay time.Duration
+	tags                    []string
+	aws                     *awslib.Client
+	metrics                 statsd.ClientInterface
+	log                     *zap.SugaredLogger
+	httpClient              *http.Client
+}
+
+type BatchMessageHandlerConfig struct {
+	APIName   string
+	JobID     string
+	QueueURL  string
+	Region    string
+	TargetURL string
+}
+
+func NewBatchMessageHandler(config BatchMessageHandlerConfig, awsClient *awslib.Client, statsdClient statsd.ClientInterface, log *zap.SugaredLogger) *BatchMessageHandler {
+	tags := []string{
+		"api_name:" + config.APIName,
+		"job_id" + config.JobID,
+	}
+
+	return &BatchMessageHandler{
+		config:                  config,
+		jobCompleteMessageDelay: _jobCompleteMessageDelay,
+		tags:                    tags,
+		aws:                     awsClient,
+		metrics:                 statsdClient,
+		log:                     log,
+		httpClient:              &http.Client{},
+	}
+}
+
+func (h *BatchMessageHandler) Handle(message *sqs.Message) error {
+	if isOnJobCompleteMessage(message) {
+		err := h.onJobComplete(message)
+		if err != nil {
+			return errors.Wrap(err, "failed to handle 'onJobComplete' message")
+		}
+		return nil
+	}
+	err := h.handleBatch(message)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (h *BatchMessageHandler) recordSuccess() error {
+	err := h.metrics.Incr("cortex_batch_succeeded", h.tags, 1.0)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	return nil
+}
+
+func (h *BatchMessageHandler) recordFailure() error {
+	err := h.metrics.Incr("cortex_batch_failed", h.tags, 1.0)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	return nil
+}
+
+func (h *BatchMessageHandler) recordTimePerBatch(elapsedTime time.Duration) error {
+	err := h.metrics.Histogram("cortex_time_per_batch", elapsedTime.Seconds(), h.tags, 1.0)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	return nil
+}
+
+func (h *BatchMessageHandler) submitRequest(messageBody string, isOnJobComplete bool) error {
+	targetURL := h.config.TargetURL
+	if isOnJobComplete {
+		targetURL = urls.Join(targetURL, "/on-job-complete")
+	}
+
+	req, err := http.NewRequest(http.MethodPost, targetURL, bytes.NewBuffer([]byte(messageBody)))
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set(CortexJobIDHeader, h.config.JobID)
+	response, err := h.httpClient.Do(req)
+	if err != nil {
+		return ErrorUserContainerNotReachable(err)
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return ErrorUserContainerResponseStatusCode(response.StatusCode)
```

Comment:
consider: if `isOnJobComplete == true` and if the error code is 404, this could be just logged as as info rather than returning an error